packages/core/core/src/Asset.js

```diff
@@ -46,63 +59,175 @@ type SerializedOptions = {|
 
 export default class Asset implements IAsset {
   id: string;
-  hash: string;
+  hash: ?string;
   filePath: FilePath;
   type: string;
-  code: string;
   ast: ?AST;
+  map: ?SourceMap;
+  mapKey: ?string;
   dependencies: Map<string, IDependency>;
   connectedFiles: Map<FilePath, File>;
-  output: AssetOutput;
+  isIsolated: boolean;
   outputHash: string;
   env: Environment;
   meta: Meta;
   stats: Stats;
+  content: Blob;
+  contentKey: ?string;
 
   constructor(options: AssetOptions) {
     this.id =
       options.id ||
       md5FromString(
         options.filePath + options.type + JSON.stringify(options.env)
       );
-    this.hash = options.hash || '';
+    this.hash = options.hash;
     this.filePath = options.filePath;
+    this.isIsolated = options.isIsolated == null ? false : options.isIsolated;
     this.type = options.type;
-    this.code = options.code || (options.output ? options.output.code : '');
+    this.content = options.content || '';
+    this.contentKey = options.contentKey;
     this.ast = options.ast || null;
+    this.map = options.map;
     this.dependencies = options.dependencies
       ? new Map(options.dependencies)
       : new Map();
     this.connectedFiles = options.connectedFiles
       ? new Map(options.connectedFiles)
       : new Map();
-    this.output = options.output || {code: this.code};
     this.outputHash = options.outputHash || '';
     this.env = options.env;
     this.meta = options.meta || {};
-    this.stats = options.stats || {
-      time: 0,
-      size: this.output.code.length
-    };
+    this.stats = options.stats;
   }
 
   serialize(): SerializedOptions {
-    // Exclude `code` and `ast` from cache
+    // Exclude `code`, `map`, and `ast` from cache
     return {
       id: this.id,
       hash: this.hash,
       filePath: this.filePath,
       type: this.type,
       dependencies: Array.from(this.dependencies),
       connectedFiles: Array.from(this.connectedFiles),
-      output: this.output,
+      isIsolated: this.isIsolated,
       outputHash: this.outputHash,
       env: this.env,
       meta: this.meta,
-      stats: this.stats
+      stats: this.stats,
+      contentKey: this.contentKey
     };
   }
 
+  /*
+   * Prepares the asset for being serialized to the cache by commiting its
+   * content and map of the asset to the cache.
+   */
+  async commit(): Promise<void> {
+    this.ast = null;
+
+    let contentStream = this.getStream();
+    if (
+      // $FlowFixMe
+      typeof contentStream.bytesRead === 'number' &&
+      contentStream.bytesRead > 0
+    ) {
+      throw new Error(
+        'Stream has already been read. This may happen if a plugin reads from a stream and does not replace it.'
+      );
+    }
+
+    let size = 0;
+    let hash = crypto.createHash('md5');
+
+    // Since we can only read from the stream once, compute the content length
+    // and hash while it's being written to the cache.
```

Comment:
I was thinking that we could work with functions that *return* readable streams, but this way we enforce that we never transform or read gigantic files unnecessarily. It does make for gymnastics like this though. But they're great for perf :)